common/archiver/gcloud/historyArchiver.go

```diff
@@ -0,0 +1,387 @@
+	}
+
+	if err := archiver.ValidateGetRequest(request); err != nil {
+		return nil, &shared.BadRequestError{Message: archiver.ErrInvalidGetHistoryRequest.Error()}
+	}
+
+	var token *getHistoryToken
+	if request.NextPageToken != nil {
+		token, err = deserializeGetHistoryToken(request.NextPageToken)
+		if err != nil {
+			return nil, &shared.BadRequestError{Message: archiver.ErrNextPageTokenCorrupted.Error()}
+		}
+	} else if request.CloseFailoverVersion != nil {
+		token = &getHistoryToken{
+			CloseFailoverVersion: *request.CloseFailoverVersion,
+			NextBatchIdx:         0,
+		}
+	} else {
+		highestVersion, historyParts, err := h.getHighestVersion(ctx, URI, request)
+		if err != nil {
+			return nil, &shared.InternalServiceError{Message: err.Error()}
+		}
+		token = &getHistoryToken{
+			CloseFailoverVersion: *highestVersion,
+			HistoryParts:         historyParts,
+			NextBatchIdx:         0,
+		}
+	}
+
+	response := &archiver.GetHistoryResponse{}
+	numOfEvents := 0
+	numOfBatches := 0
+	var pageSizeCompleted bool
+	for _, partNum := range token.HistoryParts {
+		historyBatches := []*shared.History{}
+		if pageSizeCompleted {
+			break
+		}
+
+		filename := constructHistoryFilenameMultipart(request.DomainID, request.WorkflowID, request.RunID, token.CloseFailoverVersion, partNum)
+		encodedHistoryBatches, err := h.gcloudStorage.Get(ctx, URI, filename)
+
+		if err != nil {
+			return nil, &shared.InternalServiceError{Message: err.Error()}
+		}
+
+		if encodedHistoryBatches == nil {
+			return nil, &shared.InternalServiceError{Message: "Fail retrieving history file: " + URI.String() + "/" + filename}
+		}
+
+		batches, err := decodeHistoryBatches(encodedHistoryBatches)
+		if err != nil {
+			return nil, &shared.InternalServiceError{Message: err.Error()}
+		}
+		historyBatches = append(historyBatches, batches...)
+
+		historyBatches = historyBatches[token.NextBatchIdx:]
+
+		for _, batch := range historyBatches {
+			response.HistoryBatches = append(response.HistoryBatches, batch)
+			numOfBatches++
+			numOfEvents += len(batch.Events)
+			if numOfEvents >= request.PageSize {
+				pageSizeCompleted = true
+				break
+			}
+		}
+
+		if numOfBatches < len(historyBatches) {
+			token.NextBatchIdx += numOfBatches
+			nextToken, err := serializeToken(token)
+			if err != nil {
+				return nil, &shared.InternalServiceError{Message: err.Error()}
+			}
+			response.NextPageToken = nextToken
+		}
+	}
+
+	return response, nil
+
+}
+
+// ValidateURI is used to define what a valid URI for an implementation is.
+func (h *historyArchiverData) ValidateURI(URI archiver.URI) (err error) {
+	if URI.Scheme() != URIScheme {
+		return archiver.ErrURISchemeMismatch
+	}
+
+	if URI.Path() == "" || URI.Hostname() == "" {
+		return archiver.ErrInvalidURI
+	}
+
+	return
+}
+
+func (h *historyArchiverData) validateURI(ctx context.Context, URI archiver.URI) (err error) {
+	if err = h.ValidateURI(URI); err == nil {
+		_, err = h.gcloudStorage.Exist(ctx, URI, "")
+	}
```

Comment:
Or export this function.

```
if err := h.ValidateURI(URI); err != nil {
    return err
}
return h.gcloudStorage.Exist()
```

Reply:
exist method requires the current context.

Reply:
done!